### PR TITLE
Footprints plugin

### PR DESCRIPTION
**Pull request by kecnry**
### Description

**IMPORTANT**: this PR currently require `pysiaf` to be installed or the plugin is disabled.  We may want to either remove this dependency entirely or make it an official dependency that is installed with jdaviz.

This pull request implements a new plugin in Imviz to plot an arbitrary number of user-configured footprints from any JWST instrument, adapted from code written by @melanieclarke in https://github.com/spacetelescope/jwst_novt.  In the future, this could easily be extended to other instruments or to a generic "overlays" or "regions" plugin.

Basic support for adding and customizing any number of footprints (**NOTE** some names have been changed to be more general with "overlay"/"preset" instead of "footprint"/"instrument"):

https://github.com/spacetelescope/jdaviz/assets/877591/02527a9c-4bb0-44a8-a9e6-a3719685e74d

Logic for handling multiple viewers:

https://github.com/spacetelescope/jdaviz/assets/877591/8cb7abdb-2e14-436f-ac2d-05b04f222b5b

Handling lack of WCS and ability to recenter footprint:

https://github.com/spacetelescope/jdaviz/assets/877591/c49a8970-9940-4740-a6e7-5a07db7b615c

[Footprints plugin docs](https://jdaviz--2341.org.readthedocs.build/en/2341/imviz/plugins.html#footprints)

This PR also brings the `EditableSelectPluginComponent` (which allows the user to name/rename/add/delete an arbitrary number of footprint overlays) originally implemented in https://github.com/spacetelescope/lcviz/pull/19 upstream into jdaviz, ~as well as implementing the concept of `inapplicable_attrs` to use within the plugin user APIs (if we adopt this here, we could follow-up with improvements to plot options and other plugins where some traitlets are not always applicable)~.  https://github.com/spacetelescope/lcviz/pull/36 will then update lcviz to depend on the implementation here so that we do not have two duplicate implementations.

**TODO**
- [x] docs
- [x] test coverage
- [x] screen recordings of `inapplicable_attrs` (or strip into separate PR since no longer needed now that all instruments are generalized and support the same input arguments)
- [ ] decide what to do with pysiaf dependency
- [x] consider renaming to "overlays" or "regions" for future expansion
- [x] miri & niriss support
- [ ] finalize list of apertures (and possibly make user-selectable?), including aperture used for centering
- [x] decide whether to have NIRCam:short vs long or a separate "path" dropdown that is only applicable for NIRCam (could use #2347)
- [ ] handle change in reference data (and plan what will need to be done with #2179)
- [x] fix renaming resets overlay color
- [x] rebase and confirm no issues with plot options histogram being laggy (see https://github.com/spacetelescope/jdaviz/pull/2341#issuecomment-1668452903) 

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by melanieclarke** [comment]
This looks great!  I really like being able to add and configure as many footprints as I want, from all the instruments.

A couple of picky details:
- NIRCam and NIRSpec are labeled as 'NIRcam' and 'NIRspec' - the capitalization should be fixed
- in the jwst_footprints code, there are variables named 'nrc_', which looks like a holdover from my code that treated NIRSpec and NIRCam separately.  The 'nrc' was short for NIRCam.  The variable names might be less confusing without this prefix.

And a couple comments:
- The footprints plugin seems to interact badly with the plot options histogram, particularly when the "Limit histogram to current zoom limits" option is on.  Updates for the histogram are very slow on zooming and the footprints flash on and off repeatedly while waiting.
- I can't seem to get the footprints to show up in my particular use case in the NOVT application. I'm not sure if it's my custom configuration or layout or something else.  I'll keep checking.

**Comment by kecnry** [comment]
Thanks for the feedback @melanieclarke 

> NIRCam and NIRSpec are labeled as 'NIRcam' and 'NIRspec' - the capitalization should be fixed

Fixed

> in the jwst_footprints code, there are variables named 'nrc_', which looks like a holdover from my code that treated NIRSpec and NIRCam separately. The 'nrc' was short for NIRCam. The variable names might be less confusing without this prefix.

Ah yes, I meant to clean that up after generalizing to all instruments and then forgot.  Should be fixed now too, thanks!

> The footprints plugin seems to interact badly with the plot options histogram, particularly when the "Limit histogram to current zoom limits" option is on. Updates for the histogram are very slow on zooming and the footprints flash on and off repeatedly while waiting.

I think this will be fixed by rebasing on top of #2326 - I'll rebase soon after I get test coverage up and confirm that is the case.

> I can't seem to get the footprints to show up in my particular use case in the NOVT application. I'm not sure if it's my custom configuration or layout or something else. I'll keep checking.

Ok, let me know if there's anything I can do to help!

**Comment by melanieclarke** [comment]
@kecnry - Perfect, thanks!

For the NOVT display, it looks like it's a configuration issue. If I use the Imviz configuration without any customization, the footprints work fine. My footprints appear right on top of yours, with the same parameters, as they should!  I'll sort out the config issue on my side.

**Comment by kecnry** [comment]
Few open questions to still consider during review:
* consider renaming plugin to "overlays" or "regions" for future expansion ("From File..." or API import of regions support, etc)
* decide what to do with pysiaf dependency (currently does not get installed with jdaviz, but plugin will be disabled until it is manually installed)
* finalize list of apertures (and possibly make user-selectable?), including aperture used for centering
* decide whether to have NIRCam:short vs long or a separate "path" dropdown that is only applicable for NIRCam (could use https://github.com/spacetelescope/jdaviz/pull/2347)
* follow-up tickets for any work needed when #2179 is rebased on top of this (footprint WCS-mapping will need to be updated on change to reference data)

**Comment by pllim** [comment]
I thought we agreed to bundle footprints data files and not rely on pysiaf at all? All we need is the vector points for polygons, no?

**Comment by kecnry** [comment]
> I thought we agreed to bundle footprints data files and not rely on pysiaf at all? All we need is the vector points for polygons, no?

We would also need a replacement for `pysiaf.utils.rotations.attitude` if we are to move the data locally and try to remove `pysiaf` as a dependency entirely.

**Comment by pllim** [comment]
Just these 12 lines? I'd rather we bundle this and throw in the pysiaf license here as acknowledgment than requiring the whole package.

```python
    v2d = v2 / 3600.0
    v3d = v3 / 3600.0

    # Get separate rotation matrices
    mv2 = rotate(3, -v2d)
    mv3 = rotate(2, v3d)
    mra = rotate(3, ra)
    mdec = rotate(2, -dec)
    mpa = rotate(1, -pa)

    # Combine as mra*mdec*mpa*mv3*mv2
    m = np.dot(mv3, mv2)
    m = np.dot(mpa, m)
    m = np.dot(mdec, m)
    m = np.dot(mra, m)

    return m
```

p.s. Do they mean "altitude" instead of "attitude"?

**Comment by kecnry** [comment]
Those 12 and 12 more from rotate:

```
def rotate(axis, angle):
    """Implement fundamental 3D rotation matrices.

    Rotate a vector by an angle measured in degrees, about axis 1 2 or 3 in the inertial frame.
    This is an anti-clockwise rotation when sighted along the axis, commonly called a
    right-handed rotation.

    Parameters
    ----------
    axis : int
            axis number, 1, 2, or 3
    angle : float
            angle of rotation in degrees

    Returns
    -------
    r : float array
        a (3 x 3) matrix which performs the specified rotation.

    """
    assert axis in list(range(1, 4)), 'Axis must be in range 1 to 3'
    r = np.zeros((3, 3))
    theta = np.radians(angle)

    ax0 = axis-1  # Allow for zero offset numbering
    ax1 = (ax0+1) % 3  # Axes in cyclic order
    ax2 = (ax0+2) % 3
    r[ax0, ax0] = 1.0
    r[ax1, ax1] = np.cos(theta)
    r[ax2, ax2] = np.cos(theta)
    r[ax1, ax2] = -np.sin(theta)
    r[ax2, ax1] = np.sin(theta)

    return r
```

Ok, I will look into copying everything we need to be built directly into our source and include any proper acknowledgements.

(attitude is a valid term for telescopes, I believe its used correctly)

**Comment by pllim** [comment]
I think we can stick all those in a module here and throw in the acknowledgement/license on top of that file. See https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/configs/imviz/wcs_utils.py and https://github.com/spacetelescope/jdaviz/tree/main/licenses .

Yes, you are right about attitude. Thanks for the clarification!

**Comment by kecnry** [comment]
We've discussed offline and decided to continue ahead with this review with pysiaf as a "soft-dependency" (required for the plugin, but not for jdaviz) and will consider the path forward as part of a follow-up effort. [:cat:](https://jira.stsci.edu/browse/JDAT-3697).

**Comment by rosteen** [comment]
> This is the ping-alive acting up again (sigh). Does that only happen for you when the plugin is in a popout?
> 
> EDIT: and can you re-try with the latest commit which should hopefully minimize calculations during and throttling.

I was also able to trigger without popping out, I'll pull that commit down and try again.

**Comment by rosteen** [comment]
I was unable to reproduce this with your most recent commit, seems to have fixed it (or at least made performance good enough to not trigger it on my machine).

**Comment by rosteen** [comment]
Performance/behavior/user experience all looks good to me now, I'll try to take a look at the actual code tomorrow morning.

**Review by javerbukh — APPROVED** [review]
Agreed with @rosteen that performance was great! I tested with canvas rotation, plot options, subsets, markers, etc., and never noticed unexpected delays. Docs and good look good too. Happy to retest if other reviewers ask for changes but I think this is a great start for the plugin!

**Review by rosteen — APPROVED** [review]
LGTM. There are a couple places that I wonder if we should implement one higher layer of organization for telescope (not just instrument) now in case we end up supporting e.g. HST instrument footprints, but I think we can cross that bridge if and when we come to it.

**Comment by kecnry** [comment]
I'm going to try to fix the non-WCS case, so setting back to draft to avoid accidental merges. Thanks for the reviews!

**Comment by kecnry** [comment]
Added a few commits to address the non-WCS concern by @pllim -  these make it so that the viewer entries only show in the dropdown if the viewer’s reference data has WCS, shows a warning when there are no eligible viewers, and allows for recentering the footprint on any selected viewer manually (since the defaults are originally set - changing or introducing reference data with WCS really benefits from the ability to manually recenter). 

https://github.com/spacetelescope/jdaviz/assets/877591/ac73199b-c37f-4942-98ae-9cdae9b6b077

File: jdaviz/core/template_mixin.py [hunk 1]
```diff
@@ -1403,6 +1601,13 @@ def _selected_changed(self, event):
                     return
         return super()._selected_changed(event)
 
+    def add_filter(self, *filters):
+        super().add_filter(*filters)
+        if 'reference_has_wcs' in filters:
+            # reference data can change whenever data is added OR removed from a viewer
+            self.hub.subscribe(self, AddDataMessage, handler=self._on_viewers_changed)
+            self.hub.subscribe(self, RemoveDataMessage, handler=self._on_viewers_changed)
```
Comment (kecnry) [inline jdaviz/core/template_mixin.py:1609]: @bmorris3 - this should be updated if there is an event for changing reference data in #2179

**Comment by kecnry** [comment]
after discussion, the footprint functionality is now disabled if the app is pixel-linked - we may try to remove this limitation in the future, but need to give careful thought to the expected behavior when the reference and top layer have different WCS.

This is now the current default behavior when opening the plugin (before manually overriding the default linking from pixel to WCS).


https://github.com/spacetelescope/jdaviz/assets/877591/674e600c-683c-4418-8bfc-cfcb549546b7

**Comment by javerbukh** [comment]
@kecnry Works well after testing with the latest commits and I like the paradigm you are using to ensure the WCS is available, approval still stands

**Comment by pllim** [comment]
I just want to state for the record that in a simple use case where people blink through the images so that only one is visible at a given time, the linking and reference image should not matter. All that matters is the WCS of image being viewed and the footprint. We discussed that at tag-up but agreed to defer for now.

**Comment by rosteen** [comment]
The footprints shouldn't care that linking is set to Pixel if there's only one dataset loaded in the viewer, right? Right now the footprints plugin remains disabled even if I unload all but one image from the viewer:

Screenshot 2023-08-14 at 8 58 28 AM

**Comment by kecnry** [comment]
> The footprints shouldn't care that linking is set to Pixel if there's only one dataset loaded in the viewer, right?

That's a good point - right now I do this logic based on how many are in the app's data collection.  Alternatively I guess we could have the viewer dropdown to filter based on the number loaded and the linking type, but maybe it makes more sense to push that down the road to the work to better support the pixel-linked case (which in turn is waiting on the rotation work to get merged which should simplify this greatly and probably would mean removing a lot of these guardrails anyways).  If you're ok with that, I'll add a comment to that ticket [:cat:](https://jira.stsci.edu/browse/JDAT-3699) so that we can consider this case as well.

**Comment by rosteen** [comment]
If we're potentially removing guardrails later then I'd be fine with this case being future work, as long as you add the comment to the appropriate ticket. In that case, consider my already green review to be extra green.

**Comment by kecnry** [comment]
Yes, I think we definitely want to support the pixel-linking case better in the future, its just really not clear how to do so now in a way that will still be compatible with the ongoing rotation work, so I think for now very conservative guardrails make sense so that we can move forward with other planned work for this plugin, and I've added your and @pllim's comments so we make sure we support all (possible) use-cases for pixel-linking in a hopefully self-consistent and intuitive way down the road. 

Thanks everyone for the very useful thoughts and reviews!

**Pull request merged**